### PR TITLE
Adding ordering in Before hook

### PR DESCRIPTION
**Pull request by VidhiRambhia**

### CODE REVIEW

File: lib/registry.ts [hunk 1]
```diff
@@ -291,7 +292,18 @@ export class Registry {
   }
 
   public resolveBeforeHooks(tags: string[]) {
-    return this.resolveCaseHooks("Before", tags);
+    return this.resolveCaseHooks("Before", tags).sort(
+      (a: ICaseHook, b: ICaseHook) => {
+        // If order is not specified, the hook will be executed at the end
+        if (a.order === undefined) {
+          a.order = Number.MAX_SAFE_INTEGER;
```
Comment (badeball) [inline lib/registry.ts]: Default order should be made to 10000 in order to mimic behavior seen in other implementations, ref. https://github.com/cucumber/cucumber-jvm/commit/1a5e98db62ed11c9ad151dd81219b8782e27b3f8.

File: lib/registry.ts [hunk 2]
```diff
@@ -65,6 +65,7 @@ export interface ICaseHook {
   position?: Position;
   tags?: string;
   name?: string;
+  order?: number;
```
Comment (badeball) [inline lib/registry.ts]: Make this required and give it a default value during assignment.

**Comment by badeball** [comment]
Order is an option in all hooks in the jvm implementations and should such be added here given the goal is conformity.

**Comment by badeball** [comment]
Each hook deserve more comphrehensive testing than this and requires separate tests (ie. don't mix it with "the big one"). I imagine essentially three scenarios for each hook:

- Before / Before: in order
- Before / Before(order: 10000): in order
- Before / Before(order: 9999): reversed

.. with similar tests for each other type of hook. This can be accomplished using X scenario outlines.

**Comment by VidhiRambhia** [comment]
> Order is an option in all hooks in the jvm implementations and should such be added here given the goal is conformity.

Thanks for your feedback, @badeball! So ordering has to be implemented in all of these -   `Before,
  After,
  BeforeStep,
  AfterStep,
  BeforeAll,
  AfterAll`?

Also, I understand we reverse the order of After hooks 
`  public resolveAfterHooks(tags: string[]) {
    return this.resolveCaseHooks("After", tags).reverse();
  }`
With the new approach which takes into account ordering, should this stay the same? ie we reverse sorted hooks?

**Comment by badeball** [comment]
Yes, all hooks and yes, reverse sorted in all after-type hooks!

**Comment by VidhiRambhia** [comment]
> Each hook deserve more comphrehensive testing than this and requires separate tests (ie. don't mix it with "the big one"). I imagine essentially three scenarios for each hook:
> 
> * Before / Before: in order
> * Before / Before(order: 10000): in order
> * Before / Before(order: 9999): reversed
> 
> .. with similar tests for each other type of hook. This can be accomplished using X scenario outlines.

Is it a good idea to put all of these tests in the same feature file (hooks_ordering.feature)? Or should I create a separate feature file for each hook?

**Comment by badeball** [comment]
> Is it a good idea to put all of these tests in the same feature file (hooks_ordering.feature)?

Yeah, one feature file is fine for this purpose.

File: lib/registry.ts (hunk 2 above)
Comment (VidhiRambhia) [reply lib/registry.ts]: Makes sense, addressed this

File: lib/registry.ts (hunk 1 above)
Comment (VidhiRambhia) [reply lib/registry.ts]: Done

**Comment by VidhiRambhia** [comment]
Addressed your comments @badeball. Please let me know your thoughts on it!

**Comment by badeball** [comment]
Excellent, thank you. I've merged and released this with v19.2.0. For reference, you can check out 1e3643a to see how I used scenario outlines to shorten everything.

**Comment by VidhiRambhia** [comment]
> Excellent, thank you. I've merged and released this with v19.2.0. For reference, you can check out [1e3643a](https://github.com/badeball/cypress-cucumber-preprocessor/commit/1e3643abd80fd853195d6cf5a50a923dce583f08) to see how I used scenario outlines to shorten everything.

I'll go through it. Thank you, @badeball!

**Pull request merged**